cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll

```diff
@@ -221,6 +235,16 @@ private predicate inheritanceConversionValueNumber(
   valueNumber(instr.getUnary()) = operand
 }
 
+private predicate congruentCopyInstructionTotalValueNumber(
+  CongruentCopyInstructionTotal instr, IRFunction irFunc, IRType type, ValueNumber memOperand,
+  ValueNumber operand
+) {
+  instr.getEnclosingIRFunction() = irFunc and
+  instr.getResultIRType() = type and
+  valueNumber(instr.getAnOperand().(MemoryOperand).getAnyDef()) = memOperand and
+  valueNumberOfOperand(instr.getAnOperand().(AddressOperand)) = operand
```

Comment:
What `CopyInstruction` types actually have an `AddressOperand`? I believe this PR is intended for `LoadInstruction`s, but the implementation includes all `CopyInstruction`s.
* The `CopyValueInstruction` doesn't have an `AddressOperand`, so it'll be missing from this predicate even though it's present in `numberableInstruction`, which means it'll have no value number at all, not even a unique one. I'm guessing that's not a problem in practice since `CopyValueInstruction`s read from register operands rather than memory operands, and so they must always exactly overlap.
* The `StoreInstruction` has an `AddressOperand`, but it's used for a different purpose. It doesn't have a `MemoryOperand`, though, so it'll be missing from this predicate. That's okay in practice for the same reasons as `CopyValueInstruction`.
* The QLDoc on `LoadOperand` say that this operand type is also used on `ReturnValueInstruction` and `ThrowValueInstruction`. Is there something to be done for those instructions too? I don't think there is, because those instructions don't return anything. Their loads are always exact, so they won't be relevant in this case. I wonder why those instructions have a load built into them instead of using a separate `LoadInstruction`.
* A `ReadSideEffectInstruction` is not a `CopyInstruction`, but I think it behaves like a `LoadInstruction`: it has an address and a memory operand, and that memory operand can be (is always?) inexact. It seems to me like we'd want value numbers for the indirections that are accessed by these read-side-effect instructions, but I don't see how it's possible. First, a `ReadSideEffectInstruction` does not return a result; second, we can't tell without analysing or modelling the callee what portion of the data will be loaded.

Okay, most of this discussion was just me thinking out loud and confirming that for the way the IR is generated _today_ we're fine. But I think the right thing to do, for future robustness, is to restrict `CongruentCopyInstructionTotal` to `LoadInstruction`s only and rename it accordingly.

Reply:
I've restricted it to extend `LoadInstruction` only now, and renamed it `CongruentCopyInstructionTotal` for lack of a better name. The test results are identical.